### PR TITLE
feat(lsp-ocaml): Port LS to OCaml

### PR DESCRIPTION
**Pull request by ajbt200128**
Ported the LS to OCaml!
Corresponding VSCode extension branch [here](https://github.com/returntocorp/semgrep-vscode/tree/austin/lsp-ocaml)

This should be backwards compatible with the current VSCode extension, but some of its settings are no longer relevant and don't do anything. If you're reviewing/testing this, I'd recommend checking out the branch above, and using that.

This is my first real experience with OCaml so feel free to leave any nitpicking/style guide comments!

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by aryx** [comment]
+1,861 −1,848
pretty stable change in LOC :) I guess OCaml/Python have similar expressivity

**Comment by aryx** [comment]
I think ls is not a super well know abbreviation. LSP is, so maybe rename to lsp/ or
the more verbose language_server/.
same for the package name, so we would use semgrep.language_server

Note: the LSP_experiment.ml I wrote a long time ago was to leverage type information in the OCaml language server for better typed metavariable in Semgrep for OCaml.

**Comment by ajbt200128** [comment]
> I think ls is not a super well know abbreviation. LSP is, so maybe rename to lsp/ or the more verbose language_server/. same for the package name, so we would use semgrep.language_server

Yea fair, I tried looking around at one point and some people say "LSP implementation" "LSP" or just "Language Server". The idea of calling something the "Language Server Protocol Server" makes me sad, but I've accepted it. I'll probably go for "language_server"

**Comment by aryx** [comment]
you did a mistake with git, because now we have all the changes of Brandon and Tom in your PR.
Can you merge develop correctly so that we just see your changes?

File: changelog.d/pa-lsp.added
Comment (aryx) [inline changelog.d/pa-lsp.added:1]: ... has been reimplemented in OCaml ...

File: cli/src/semgrep/lsp/schema.json
Comment (aryx) [inline cli/src/semgrep/lsp/schema.json:81]: what happened to all those schema elements? We don't support those features anymore?

File: cli/src/semgrep/metrics.py
Comment (aryx) [inline cli/src/semgrep/metrics.py:194]: Why we don't care about fix_rate anymore?

File: src/ls/CodeActions.mli
Comment (aryx) [inline src/ls/CodeActions.mli:1]: In general we try very hard to document things in the .mli. In fact, most of the time for PRs I just look at the .mli.
So here it would be probably better to rename the function to code_actions_of_core_matches
and explain what is the string list second argument (since it's not super clear from the type what it is).

File: src/ls/CodeActions.ml
Comment (aryx) [inline src/ls/CodeActions.ml:40]: Use of Common2 is deprecated. Here you can probably use List.concat_map
We could add a semgrep rule for it  ...

File: src/ls/CodeActions.mli
Comment (aryx) [reply src/ls/CodeActions.mli:1]: I guess the string list is a list of filenames. In theory you could use Common.filename, which is an alias to a string,
but recently we started to switch to more strongly typed file names with Fpath.t.
That requires to use a few extra functions. See also libs/commons/File.mli

File: src/ls/CodeActions.ml
Comment (aryx) [inline src/ls/CodeActions.ml:4]: A Toplevel comment in the file rougly explaining the goal of the module would be helpful.
See other files in semgrep as example (e.g., AST_generic.ml, but does not have have to be that long :) )

File: src/ls/Diagnostics.mli
Comment (aryx) [inline src/ls/Diagnostics.mli:1]: Again, a small comment in the .mli will help.
The .mli are the most important files.
That's usually what we read when we want to use other people's code.

File: src/ls/Git_helper.ml
Comment (aryx) [inline src/ls/Git_helper.ml:1]: This is mostly independent of the language server, so it should be moved out, e.g., in libs/
and actually there are a few places where we do git related stuff that we should merge together.
maybe we can start a libs/lib_git/ to gather such code.

File: src/ls/Git_helper.ml
Comment (aryx) [inline src/ls/Git_helper.ml:19]: @dinosaure started to use the Bos library for such things. We should probably unify things.

File: src/ls/LS.mli
Comment (aryx) [inline src/ls/LS.mli:1]: comment

Comment (aryx) [reply src/ls/LS.mli:1]: Can be "Entry point of the language server. This will start the server following the LSP protocol and listen on port XXX for requests from VScode or other LSP clients.

File: src/ls/Lsp_util.mli
Comment (aryx) [inline src/ls/Lsp_util.mli:1]: Those are actually not utilities related to LSP but utilities related to Semgrep.
Maybe Convert_utils.mli

File: src/ls/Lsp_util.mli
Comment (aryx) [inline src/ls/Lsp_util.mli:2]: This does not return a strings maybe lsp_severity_of_rule_severity
(the final type may be a string, but this is an implementation detail we should hide. You use DiagnosticSeverity.Error
in the code, so this looks like a real type)

**Review by aryx — APPROVED** [review]

File: src/ls/Processed_run.ml
Comment (aryx) [inline src/ls/Processed_run.ml:7]: Ok, I'll try to fix the dependency cycle.

File: src/ls/Processed_run.mli
Comment (aryx) [inline src/ls/Processed_run.mli:3]: small comment, again. Anything that can help the consumer of this module to understand how to use it.

File: src/ls/Unit_LS.ml
Comment (aryx) [inline src/ls/Unit_LS.ml:6]: great comment

File: src/ls/Unit_LS.ml
Comment (aryx) [inline src/ls/Unit_LS.ml:143]: @mjambon started to write some test utilities in help creating test files and dirs.
See libs/Testutil_files.mli

Comment (aryx) [reply src/ls/Unit_LS.ml:143]: Maybe we need to extend it to support those ~git:xxx flags.

File: src/ls/dune
Comment (aryx) [inline src/ls/dune:2]: I think ls -> language_server everywhere will be better.

**Review by aryx — APPROVED** [review]
Great work!

**Comment by aryx** [comment]
Usually when we approve but don't merge the PR that means the PR is mostly good, but we let the author
address the comments and then merge it. There is no need for a second review, just try your best to address the comment
(if you disagree then don't change the code).

**Comment by aryx** [comment]
The CI failure seems unrelated. I'll try to fix it. All the PRs on develop are currently broken because of this dev-docker CI step failing.

File: src/ls/Git_helper.ml
Comment (dinosaure) [reply src/ls/Git_helper.ml:19]: I think it's indeed the right choice & allow us to be outside the `lwt` monad.

**Pull request merged**
